src/main/java/io/jenkins/plugins/coverage/QualityGateEvaluator.java

```diff
@@ -0,0 +1,66 @@
+package io.jenkins.plugins.coverage;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.google.errorprone.annotations.FormatMethod;
+
+import io.jenkins.plugins.coverage.model.CoverageNode;
+
+/**
+ * Evaluates a set of quality gates for a static analysis report.
+ *
+ * @author Michael Müller, Nikolas Paripovic
+ */
+public class QualityGateEvaluator {
+
+    /**
+     * Enforces this quality gate for the specified run.
+     *
+     * @param coverageNode
+     *         the coverage root node to evaluate
+     * @param qualityGates
+     *          the quality gates to be checked
+     * @param logger
+     *         the logger that reports the passed and failed quality gate thresholds
+     *
+     * @return result of the evaluation, expressed by a build state
+     */
+    public QualityGateStatus evaluate(final CoverageNode coverageNode, final List<QualityGate> qualityGates,
+            final FormattedLogger logger) {
+
+        if (qualityGates.isEmpty()) {
+            logger.print("-> INACTIVE - No quality gate defined");
+            return QualityGateStatus.INACTIVE;
+        }
+
+        return qualityGates.stream().reduce(QualityGateStatus.PASSED,
+                (currentStatus, qualityGate) ->
+                    Optional.ofNullable(coverageNode.getCoverage(qualityGate.getCoverageMetric()))
+                            .filter(coverage -> coverage.getCoveredPercentage() < qualityGate.getThreshold())
+                            .map(coverage -> qualityGate.getStatusIfNotPassedSuccessful())
+                            .filter(status -> status.isWorseThan(currentStatus))
+                            .orElse(currentStatus),
+                QualityGateStatus::getWorseStatus);
```

Comment:
This is an example where I think that the imperative style is more expressive than the functional one.